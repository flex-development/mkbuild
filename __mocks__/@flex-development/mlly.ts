/**
 * @file Mocks - @flex-development/mlly
 * @module mocks/flex-development/mlly
 * @see https://github.com/flex-development/mlly
 */

/**
 * [`@flex-development/mlly`][1] module type.
 *
 * [1]: https://github.com/flex-development/mlly
 */
type Actual = typeof import('@flex-development/mlly')

/**
 * `@flex-development/mlly` module.
 *
 * @const {Actual} actual
 */
const actual: Actual = await vi.importActual<Actual>('@flex-development/mlly')

export const CONDITIONS = actual.CONDITIONS
export const RESOLVE_EXTENSIONS = actual.RESOLVE_EXTENSIONS
export const fillModules = vi.fn(actual.fillModules)
export const findStaticImports = vi.fn(actual.findStaticImports)
export const getSource = vi.fn(actual.getSource)
export const readPackageJson = vi.fn(actual.readPackageJson)
export const resolveModule = vi.fn(actual.resolveModule)
export const resolveModules = vi.fn(actual.resolveModules)
export const toDataURL = vi.fn(actual.toDataURL)
export const toURL = vi.fn(actual.toURL)
