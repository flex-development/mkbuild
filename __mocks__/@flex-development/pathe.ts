/**
 * @file Mocks - @flex-development/pathe
 * @module mocks/flex-development/pathe
 * @see https://github.com/flex-development/pathe
 */

/**
 * [`@flex-development/pathe`][1] module type.
 *
 * [1]: https://github.com/flex-development/pathe
 */
type Actual = typeof import('@flex-development/pathe')

/**
 * `@flex-development/pathe` module.
 *
 * @const {Actual} actual
 */
const actual: Actual = await vi.importActual<Actual>('@flex-development/pathe')

export const basename = vi.fn(actual.basename)
export const dirname = vi.fn(actual.dirname)
export const extname = vi.fn(actual.extname)
export const format = vi.fn(actual.format)
export const join = vi.fn(actual.join)
export const parse = vi.fn(actual.parse)
export const relative = vi.fn(actual.relative)
export const resolve = vi.fn(actual.resolve)

export default {
  basename,
  dirname,
  extname,
  format,
  join,
  parse,
  relative,
  resolve
}
