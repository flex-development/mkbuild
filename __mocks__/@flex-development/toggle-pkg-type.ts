/**
 * @file Mocks - @flex-development/toggle-pkg-type
 * @module mocks/flex-development/toggle-pkg-type
 * @see https://github.com/flex-development/toggle-pkg-type
 */

export default vi.fn()
