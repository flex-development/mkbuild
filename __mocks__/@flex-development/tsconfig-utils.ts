/**
 * @file Mocks - @flex-development/tsconfig-utils
 * @module mocks/flex-development/tsconfig-utils
 * @see https://github.com/flex-development/tsconfig-utils
 */

/**
 * [`@flex-development/tsconfig-utils`][1] module type.
 *
 * [1]: https://github.com/flex-development/tsconfig-utils
 */
type Actual = typeof import('@flex-development/tsconfig-utils')

/**
 * `@flex-development/tsconfig-utils` module.
 *
 * @const {Actual} actual
 */
const actual: Actual = await vi.importActual<Actual>(
  '@flex-development/tsconfig-utils'
)

export const ScriptTarget = actual.ScriptTarget
export const normalizeCompilerOptions = vi.fn(actual.normalizeCompilerOptions)
export const normalizeModuleResolution = vi.fn(actual.normalizeModuleResolution)
export const resolvePaths = vi.fn(actual.resolvePaths)
