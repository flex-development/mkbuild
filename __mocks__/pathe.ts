/**
 * @file Mocks - pathe
 * @module mocks/pathe
 * @see https://github.com/unjs/pathe
 */

import path, { type FormatInputPathObject } from 'node:path'

/**
 * Return the last portion of `p`.
 *
 * @param {string} p - Path to evaluate
 * @return {string} File name
 */
export const basename = vi.fn((p: string): string => path.basename(p))

/**
 * Returns the directory name of `p`.
 *
 * @param {string} p - Path to evaluate
 * @return {string} Directory containing `p`
 */
export const dirname = vi.fn((p: string): string => path.dirname(p))

/**
 * Returns the extension of `p`, from the last `'.'` to end of string in the
 * last portion of the path.
 *
 * If there is no `'.'` in the last portion of the path or the first character
 * of it is `'.'`, an empty string will be returned.
 *
 * @param {string} p - Path to evaluate
 * @return {string} Extension of `p`
 */
export const extname = vi.fn((p: string) => path.extname(p))

/**
 * Returns a path string from an object - the opposite of {@link parse}.
 *
 * @param {FormatInputPathObject} p - Path object to evaluate
 * @return {path.ParsedPath} Object representing `p` parsed
 */
export const format = vi.fn((p: FormatInputPathObject) => path.format(p))

/**
 * Parses `p`.
 *
 * @param {string} p - Path to evaluate
 * @return {path.ParsedPath} Object representing `p` parsed
 */
export const parse = vi.fn((p: string) => path.parse(p))

/**
 * Returns the relative path from `from` to `to` based on the current working
 * directory.
 *
 * @param {string} from - Start path
 * @param {string} to - Destination path
 * @return {string} Relative path from `from` to `to`
 */
export const relative = vi.fn((from: string, to: string): string => {
  return path.relative(from, to)
})

/**
 * Starting from leftmost {from} parameter, resolves {to} to an absolute path.
 *
 * @param {string[]} paths - Path segments
 * @return {string} Absolute path composed from `paths`
 */
export const resolve = vi.fn((...paths: string[]): string => {
  return path.resolve(...paths)
})
