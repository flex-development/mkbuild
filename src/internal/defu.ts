/**
 * @file Internal - defu
 * @module mkbuild/internal/defu
 */

import {
  isUndefined,
  type ExactOptionalPropertyTypes,
  type ObjectPlain
} from '@flex-development/tutils'
import { mergeAndCompare as merge, type Merge } from 'merge-anything'

/**
 * Assigns default properties to the given `object`.
 *
 * Leftmost arguments have more priority when assigning defaults.
 *
 * @template T - Destination object type
 * @template D - Source object type(s)
 *
 * @param {T} object - Destination object
 * @param {D} defaults - Object(s) used to assign default properties
 * @return {ExactOptionalPropertyTypes<Merge<T, D> & T>} `object` with defaults
 */
function defu<T extends ObjectPlain, D extends ObjectPlain[]>(
  object: T,
  ...defaults: D
): ExactOptionalPropertyTypes<Merge<T, D> & T> {
  return merge<T, D>(
    /**
     * Determines how to assign a property.
     *
     * @param {unknown} v1 - Original value from {@linkcode object}
     * @param {unknown} v2 - Incoming value from object in {@linkcode defaults}
     * @return {unknown} Property value for {@linkcode object}
     */
    (v1: unknown, v2: unknown): unknown => (isUndefined(v1) ? v2 : v1),
    object,
    ...defaults
  )
}

export default defu
