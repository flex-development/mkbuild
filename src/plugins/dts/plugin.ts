/**
 * @file Plugins - dts
 * @module mkbuild/plugins/dts
 */

import defu from '#src/internal/defu'
import type { OutputMetadata } from '#src/types'
import {
  EXT_DTS_REGEX,
  EXT_JS_REGEX,
  EXT_TS_REGEX
} from '@flex-development/ext-regex'
import pathe from '@flex-development/pathe'
import * as tscu from '@flex-development/tsconfig-utils'
import type {
  BuildOptions,
  BuildResult,
  OnResolveArgs,
  OutputFile,
  Plugin,
  PluginBuild
} from 'esbuild'
import util from 'node:util'
import type {
  CompilerHost,
  CompilerOptions,
  WriteFileCallback
} from 'typescript'

/**
 * Returns a TypeScript declaration plugin.
 *
 * @return {Plugin} TypeScript declaration plugin
 */
const plugin = (): Plugin => {
  /**
   * Generates TypeScript declarations.
   *
   * @todo handle diagnostics
   * @todo bundle declarations
   *
   * [1]: https://esbuild.github.io/plugins
   * [2]: https://esbuild.github.io/api/#build-api
   *
   * @async
   *
   * @param {PluginBuild} build - [esbuild plugin api][1]
   * @param {BuildOptions} build.initialOptions - [esbuild build api][2] options
   * @param {PluginBuild['onEnd']} build.onEnd - Build end cb
   * @param {PluginBuild['onResolve']} build.onResolve - Path resolution cb
   * @return {Promise<void>} Nothing when complete
   * @throws {Error}
   */
  const setup = async ({
    initialOptions,
    onEnd,
    onResolve
  }: PluginBuild): Promise<void> => {
    const {
      absWorkingDir = process.cwd(),
      bundle,
      color = true,
      format,
      metafile,
      outExtension: { '.js': ext = '.js' } = {},
      outbase = '',
      outdir = '.',
      preserveSymlinks = false,
      tsconfig = 'tsconfig.json',
      write
    } = initialOptions

    // not bundling declarations yet
    if (bundle) return void bundle

    // esbuild write must be disabled to access result.outputFiles
    if (write) throw new Error('write must be disabled')

    // metafile required to get output metadata
    if (!metafile) throw new Error('metafile required')

    /**
     * TypeScript module.
     *
     * @const {typeof import('typescript')} ts
     */
    const ts: typeof import('typescript') = (await import('typescript')).default

    /**
     * Absolute paths to source files.
     *
     * @const {string[]} sourcefiles
     */
    const sourcefiles: string[] = []

    /**
     * Source file filter.
     *
     * @const {RegExp} filter
     */
    const filter: RegExp = new RegExp(
      (
        `(?:(?:${EXT_JS_REGEX.source.slice(0, -8)})` +
        `|(?:${EXT_TS_REGEX.source.slice(0, -8)}))$`
      ).replace(/\?<.+?>/g, '?:')
    )

    // get source files as entry points are resolved
    onResolve({ filter }, (args: OnResolveArgs): undefined => {
      return void sourcefiles.push(pathe.resolve(args.resolveDir, args.path))
    })

    return void onEnd((result: BuildResult): void => {
      /**
       * TypeScript compiler options.
       *
       * @var {CompilerOptions | tscu.CompilerOptions} compilerOptions
       */
      let compilerOptions: CompilerOptions | tscu.CompilerOptions =
        tscu.loadCompilerOptions(pathe.resolve(absWorkingDir, tsconfig))

      // normalize compiler options
      compilerOptions = tscu.normalizeCompilerOptions(compilerOptions)

      // remove forbidden compiler options
      delete compilerOptions.declarationDir
      delete compilerOptions.out
      delete compilerOptions.outFile
      delete compilerOptions.rootDirs
      delete compilerOptions.sourceMap
      delete compilerOptions.sourceRoot

      // remove overridden compiler options
      delete compilerOptions.declaration
      delete compilerOptions.emitDeclarationOnly
      delete compilerOptions.noEmit
      delete compilerOptions.noErrorTruncation
      delete compilerOptions.outDir
      delete compilerOptions.rootDir

      // remove unsupported compiler options
      delete compilerOptions.declarationMap
      delete compilerOptions.noEmitOnError

      // merge compiler options with defaults
      compilerOptions = defu(compilerOptions, {
        allowJs: true,
        allowUmdGlobalAccess: format === 'iife',
        allowUnreachableCode: false,
        baseUrl: absWorkingDir,
        checkJs: false,
        declaration: true,
        declarationMap: false,
        emitDeclarationOnly: true,
        esModuleInterop: true,
        forceConsistentCasingInFileNames: true,
        isolatedModules: true,
        module: ts.ModuleKind.ESNext,
        moduleResolution: tscu.normalizeModuleResolution(
          tscu.ModuleResolutionKind.NodeJs
        ),
        noEmit: false,
        noEmitOnError: false,
        noErrorTruncation: true,
        noImplicitAny: true,
        noImplicitOverride: true,
        noImplicitReturns: true,
        outDir: pathe.resolve(absWorkingDir, outdir),
        preserveConstEnums: true,
        preserveSymlinks,
        pretty: color,
        resolveJsonModule: true,
        rootDir: pathe.resolve(absWorkingDir, outbase),
        skipLibCheck: true,
        target: ts.ScriptTarget.ESNext
      })

      /**
       * TypeScript compiler host.
       *
       * @const {CompilerHost} host
       */
      const host: CompilerHost = ts.createCompilerHost(compilerOptions)

      /**
       * Virtual file system for declaration files.
       *
       * @const {Map<string, string>} vfs
       */
      const vfs: Map<string, string> = new Map()

      // first letter before "js" or "ts" in output file extension
      const [, cm = ''] = /\.(c|m)?[jt]sx?$/.exec(ext)!

      /**
       * Writes a declaration file to {@linkcode vfs}.
       *
       * @param {string} filename - Name of file to write
       * @param {string} contents - Content to write
       * @return {void} Nothing when complete
       */
      const writeFile: WriteFileCallback = (
        filename: string,
        contents: string
      ): void => {
        return void vfs.set(
          filename.replace(EXT_DTS_REGEX, `.d.${cm}ts`),
          contents
        )
      }

      // override write file function
      host.writeFile = writeFile

      // emit declarations to virtual file system
      ts.createProgram(
        sourcefiles.filter(sourcefile => !EXT_DTS_REGEX.test(sourcefile)),
        compilerOptions,
        host
      ).emit()

      // remap output files to insert declaration file outputs
      return void (result.outputFiles = result.outputFiles!.flatMap(output => {
        /**
         * Absolute path to declaration file for {@linkcode output}.
         *
         * @const {string} dtspath
         */
        const dtspath: string = EXT_JS_REGEX.test(output.path)
          ? output.path.replace(EXT_JS_REGEX, '.d.$1ts')
          : EXT_TS_REGEX.test(output.path) && !EXT_DTS_REGEX.test(output.path)
          ? output.path.replace(EXT_TS_REGEX, '.d.$2ts')
          : ''

        // do nothing if missing declaration file
        if (!vfs.has(dtspath)) return [output]

        /**
         * Relative path to output file.
         *
         * **Note**: Relative to {@linkcode absWorkingDir}.
         *
         * @const {string} outfile
         */
        const outfile: string = output.path
          .replace(absWorkingDir, '')
          .replace(/^\//, '')

        /**
         * {@linkcode output} metadata.
         *
         * @const {OutputMetadata} metadata
         */
        const metadata: OutputMetadata = result.metafile!.outputs[outfile]!

        /**
         * Declaration output file content.
         *
         * @const {string} dtstext
         */
        const dtstext: string = vfs.get(dtspath)!

        /**
         * Declaration file output.
         *
         * @const {OutputFile} dts
         */
        const dts: OutputFile = {
          contents: new util.TextEncoder().encode(dtstext),
          path: dtspath,
          text: dtstext
        }

        /**
         * Relative path to declaration output file.
         *
         * **Note**: Relative to {@linkcode absWorkingDir}.
         *
         * @const {string} dtsoutfile
         */
        const dtsoutfile: string = dtspath
          .replace(absWorkingDir, '')
          .replace(/^\//, '')

        // update metafile
        result.metafile!.outputs[dtsoutfile] = {
          ...metadata,
          bytes: Buffer.byteLength(dts.contents)
        }

        return [output, dts]
      }))
    })
  }

  return { name: 'dts', setup }
}

export default plugin
