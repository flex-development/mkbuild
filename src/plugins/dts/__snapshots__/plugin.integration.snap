// Vitest Snapshot v1

exports[`integration:plugins/dts > esbuild > should add .d.cts output 1`] = `
{
  "path": "dist/my-atoi.d.cts",
  "text": "/**
 * @file Fixtures - myAtoi
 * @module fixtures/myAtoi
 * @see https://leetcode.com/problems/string-to-integer-atoi
 */
/**
 * Converts a string, \`s\`, to a 32-bit signed integer (similar to C/C++'s \`atoi\`
 * function).
 *
 * @example myAtoi('') // 0
 * @example myAtoi('42') // 42
 * @example myAtoi('   -42') // -42
 * @example myAtoi('4193 with words') // 4193
 * @example myAtoi('words and 987') // 0
 * @example myAtoi('+-12') // 0
 *
 * @param {string} s - String to convert
 * @return {number} 32-bit signed integer
 */
declare function myAtoi(s: string): number;
export default myAtoi;
",
}
`;

exports[`integration:plugins/dts > esbuild > should add .d.cts output for copied file 1`] = `
{
  "path": "dist/sum-of-intervals.d.cts",
  "text": "export = sumOfIntervals;
/**
 * @file Fixtures - sumOfIntervals
 * @module fixtures/sum-of-intervals.cjs
 * @see https://codewars.com/kata/52b7ed099cdc285c300001cd
 */
/**
 * Given an array of intervals, all of varying lengths, the function returns the
 * sum of all interval lengths. **Overlapping intervals will be counted once**.
 *
 * @example
 *  sumOfIntervals([[1, 5]]) // 4
 * @example
 *  sumOfIntervals([[1, 4], [7, 10], [3, 5]]) // 7
 * @example
 *  sumOfIntervals([[1, 5], [10, 15], [-1, 3]]) // 11
 *
 * @param {[number, number][]} intervals - Intervals
 * @return {number} Sum of all interval lengths
 */
declare function sumOfIntervals(intervals: [number, number][]): number;
",
}
`;

exports[`integration:plugins/dts > esbuild > should add .d.mts output 1`] = `
{
  "path": "dist/reverse.d.mts",
  "text": "/**
 * @file Fixtures - reverse
 * @module fixtures/reverse
 * @see https://leetcode.com/problems/reverse-integer
 */
/**
 * Given a signed 32-bit integer, \`x\`, the function returns \`x\` with its digits
 * reversed. If reversing \`x\` causes the value to go outside the signed 32-bit
 * integer range, \`0\` will be returned instead.
 *
 * @example reverse(3) // 3
 * @example reverse(-4) // -4
 * @example reverse(123) // 321
 * @example reverse(-123) // -321
 * @example reverse(120) // 21
 * @example reverse(1534236469) // 0
 *
 * @param {number} x - Integer to reverse
 * @return {number} \`x\` with its digits reversed
 */
declare function reverse(x: number): number;
export default reverse;
",
}
`;

exports[`integration:plugins/dts > esbuild > should add .d.mts output for copied file 1`] = `
{
  "path": "dist/sum-of-intervals.d.mts",
  "text": "export default sumOfIntervals;
/**
 * @file Fixtures - sumOfIntervals
 * @module fixtures/sum-of-intervals.mjs
 * @see https://codewars.com/kata/52b7ed099cdc285c300001cd
 */
/**
 * Given an array of intervals, all of varying lengths, the function returns the
 * sum of all interval lengths. **Overlapping intervals will be counted once**.
 *
 * @example
 *  sumOfIntervals([[1, 5]]) // 4
 * @example
 *  sumOfIntervals([[1, 4], [7, 10], [3, 5]]) // 7
 * @example
 *  sumOfIntervals([[1, 5], [10, 15], [-1, 3]]) // 11
 *
 * @param {[number, number][]} intervals - Intervals
 * @return {number} Sum of all interval lengths
 */
declare function sumOfIntervals(intervals: [number, number][]): number;
",
}
`;

exports[`integration:plugins/dts > esbuild > should add .d.ts output 1`] = `
{
  "path": "dist/dbl-linear.d.ts",
  "text": "/**
 * @file Fixtures - dblLinear
 * @module fixtures/dblLinear
 * @see https://codewars.com/kata/5672682212c8ecf83e000050
 */
/**
 * Consider a sequence \`u\` where \`u\` is defined as follows:
 *
 * 1. The number \`u(0) = 1\` is the first one in \`u\`
 * 2. For each \`x\` in \`u\`, then \`y = 2x + 1\` and \`z = 3x + 1\` must be in \`u\` too
 * 3. There are no other numbers in \`u\`
 *
 * Given an index, \`n\`, the function returns the element at \`u(n)\`.
 *
 * @example
 *  dblLinear(0) // 1
 * @example
 *  dblLinear(10) // 22
 * @example
 *  dblLinear(100) // 447
 * @example
 *  dblLinear(7687) // 111718
 *
 * @param {number} n - Index of element to get
 * @return {number} Element at \`u(n)\`
 */
declare const dblLinear: (n: number) => number;
export default dblLinear;
",
}
`;

exports[`integration:plugins/dts > esbuild > should skip files that are not javascript or typescript 1`] = `
"[
  {
    \\"date\\": \\"2007-04-24T07:00:00.000Z\\",
    \\"close\\": 93.24
  },
  {
    \\"date\\": \\"2007-04-25T07:00:00.000Z\\",
    \\"close\\": 95.35
  },
  {
    \\"date\\": \\"2007-06-08T07:00:00.000Z\\",
    \\"close\\": 124.49
  },
  {
    \\"date\\": \\"2007-06-11T07:00:00.000Z\\",
    \\"close\\": 120.19
  },
  {
    \\"date\\": \\"2007-06-12T07:00:00.000Z\\",
    \\"close\\": 120.38
  }
]
"
`;
